src/team_lib/mhba/xccl_mhba_lib.c

```diff
@@ -583,46 +608,52 @@ xccl_mhba_team_destroy(xccl_tl_team_t *team)
         if (status!=XCCL_OK){
             xccl_mhba_error("failed to destroy Mkeys");
         }
+        free(mhba_team->net.rkeys);
     }
     free(team);
     return status;
 }
 
-xccl_status_t xccl_mhba_node_fanin(xccl_mhba_team_t *team, int fanin_value, int root)
+xccl_status_t xccl_mhba_node_fanin(xccl_mhba_team_t *team, xccl_mhba_coll_req_t *request)
 {
     int i;
     int *ctrl_v;
-    if (team->node.sbgp->group_rank != root) {
-        ctrl_v = (int*)team->node.my_ctrl;
-        int v = *ctrl_v;
-        __sync_fetch_and_add(ctrl_v, (v+fanin_value));
+    int index = seq_index(request->seq_num);
+
+    if(team->occupied_operations_slots[index]){
+        return XCCL_INPROGRESS;
+    } //wait for slot to be open
+    team->occupied_operations_slots[seq_index(team->sequence_number)] = 1;
+    xccl_mhba_update_mkeys_entries(&team->node, request); // no option for failure status
```

Comment:
why update mkeys is inside fanin ? this makes no sense. Fanin is a synchronizing primitive, it should not have such side effects as "updating the mkey"

Reply:
we had all that logic (check of the slot + update key) in the schedule function. Why it was moved here?

Reply:
added to do - change fainin name